File: examples/lwjgl-map/src/main/java/org/maplibre/nativeffi/examples/lwjglmap/OpenGLContext.java
package org.maplibre.nativeffi.examples.lwjglmap;

import static org.lwjgl.glfw.GLFW.GLFW_CLIENT_API;
import static org.lwjgl.glfw.GLFW.GLFW_CONTEXT_CREATION_API;
import static org.lwjgl.glfw.GLFW.GLFW_CONTEXT_VERSION_MAJOR;
import static org.lwjgl.glfw.GLFW.GLFW_CONTEXT_VERSION_MINOR;
import static org.lwjgl.glfw.GLFW.GLFW_EGL_CONTEXT_API;
import static org.lwjgl.glfw.GLFW.GLFW_OPENGL_API;
import static org.lwjgl.glfw.GLFW.GLFW_OPENGL_ES_API;
import static org.lwjgl.glfw.GLFW.GLFW_RESIZABLE;
import static org.lwjgl.glfw.GLFW.GLFW_TRUE;
import static org.lwjgl.glfw.GLFW.glfwCreateWindow;
import static org.lwjgl.glfw.GLFW.glfwDefaultWindowHints;
import static org.lwjgl.glfw.GLFW.glfwDestroyWindow;
import static org.lwjgl.glfw.GLFW.glfwGetVersionString;
import static org.lwjgl.glfw.GLFW.glfwInit;
import static org.lwjgl.glfw.GLFW.glfwMakeContextCurrent;
import static org.lwjgl.glfw.GLFW.glfwSwapBuffers;
import static org.lwjgl.glfw.GLFW.glfwTerminate;
import static org.lwjgl.glfw.GLFW.glfwWindowHint;
import static org.lwjgl.glfw.GLFWNativeEGL.glfwGetEGLConfig;
import static org.lwjgl.glfw.GLFWNativeEGL.glfwGetEGLContext;
import static org.lwjgl.glfw.GLFWNativeEGL.glfwGetEGLDisplay;
import static org.lwjgl.glfw.GLFWNativeEGL.glfwGetEGLSurface;
import static org.lwjgl.glfw.GLFWNativeWGL.glfwGetWGLContext;
import static org.lwjgl.glfw.GLFWNativeWin32.glfwGetWin32Window;
import static org.lwjgl.system.MemoryUtil.NULL;

import java.util.Locale;
import org.lwjgl.PointerBuffer;
import org.lwjgl.opengl.GL;
import org.lwjgl.opengles.GLES;
import org.lwjgl.system.MemoryStack;
import org.lwjgl.system.windows.User32;
import org.maplibre.nativeffi.Maplibre;
import org.maplibre.nativeffi.render.EglContextDescriptor;
import org.maplibre.nativeffi.render.NativePointer;
import org.maplibre.nativeffi.render.OpenGLContextDescriptor;
import org.maplibre.nativeffi.render.OpenGLContextProvider;
import org.maplibre.nativeffi.render.RenderBackend;
import org.maplibre.nativeffi.render.WglContextDescriptor;

final class OpenGLContext implements GraphicsContext {
  private final boolean gles;
  private final long window;
  private long hdc;
  private long eglDisplay;
  private long eglConfig;
  private long eglSurface;
  private long eglContext;
  private boolean closed;

  private OpenGLContext(boolean gles, long window) {
    this.gles = gles;
    this.window = window;
  }

  static OpenGLContext create(String title, int width, int height) {
    if (isLinux()) {
      return createEgl(title, width, height);
    }
    if (isWindows()) {
      return createWgl(title, width, height);
    }
    throw new IllegalStateException("LWJGL OpenGL context is only supported on Linux and Windows");
  }

  private static OpenGLContext createEgl(String title, int width, int height) {
    if (!Maplibre.supportedOpenGLContextProviders().contains(OpenGLContextProvider.EGL)) {
      throw new IllegalStateException("Native library does not support EGL");
    }
    if (!glfwInit()) {
      throw new IllegalStateException("GLFW initialization failed");
    }
    long window = NULL;
    var context = new OpenGLContext(true, NULL);
    try (var stack = MemoryStack.stackPush()) {
      glfwDefaultWindowHints();
      glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_ES_API);
      glfwWindowHint(GLFW_CONTEXT_CREATION_API, GLFW_EGL_CONTEXT_API);
      glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
      glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
      glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE);
      window = glfwCreateWindow(width, height, title, NULL, NULL);
      if (window == NULL) {
        throw new IllegalStateException("GLFW EGL window creation failed");
      }
      context = new OpenGLContext(true, window);
      context.makeCurrent();
      GLES.createCapabilities();
      context.eglDisplay = glfwGetEGLDisplay();
      context.eglContext = glfwGetEGLContext(window);
      context.eglSurface = glfwGetEGLSurface(window);
      PointerBuffer configs = stack.mallocPointer(1);
      if (!glfwGetEGLConfig(window, configs)) {
        throw new IllegalStateException("GLFW did not expose an EGL config");
      }
      context.eglConfig = configs.get(0);
      if (context.eglDisplay == NULL || context.eglContext == NULL || context.eglSurface == NULL) {
        throw new IllegalStateException("GLFW did not expose EGL handles");
      }
      System.out.printf("GLFW %s, OpenGL EGL/GLES%n", glfwGetVersionString());
      return context;
    } catch (RuntimeException error) {
      context.close();
      if (window != NULL && context.window != window) {
        glfwDestroyWindow(window);
        glfwTerminate();
      }
      throw error;
    }
  }

  private static OpenGLContext createWgl(String title, int width, int height) {
    if (!Maplibre.supportedOpenGLContextProviders().contains(OpenGLContextProvider.WGL)) {
      throw new IllegalStateException("Native library does not support WGL");
    }
    if (!glfwInit()) {
      throw new IllegalStateException("GLFW initialization failed");
    }
    var context = new OpenGLContext(false, NULL);
    try {
      glfwDefaultWindowHints();
      glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_API);
      glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
      glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
      glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE);
      var window = glfwCreateWindow(width, height, title, NULL, NULL);
      if (window == NULL) {
        throw new IllegalStateException("GLFW WGL window creation failed");
      }
      context = new OpenGLContext(false, window);
      context.makeCurrent();
      GL.createCapabilities();
      var hwnd = glfwGetWin32Window(window);
      var hglrc = glfwGetWGLContext(window);
      context.hdc = getDc(hwnd);
      if (context.hdc == NULL || hglrc == NULL) {
        throw new IllegalStateException("GLFW did not expose WGL handles");
      }
      System.out.printf("GLFW %s, OpenGL WGL%n", glfwGetVersionString());
      return context;
    } catch (RuntimeException error) {
      context.close();
      throw error;
    }
  }

  @Override
  public long window() {
    return window;
  }

  @Override
  public RenderBackend backend() {
    return RenderBackend.OPENGL;
  }

  boolean isGles() {
    return gles;
  }

  OpenGLContextDescriptor descriptor() {
    if (gles) {
      return new EglContextDescriptor(
          NativePointer.ofAddress(eglDisplay),
          NativePointer.ofAddress(eglConfig),
          NativePointer.ofAddress(eglContext));
    }
    return new WglContextDescriptor(
        NativePointer.ofAddress(hdc), NativePointer.ofAddress(glfwGetWGLContext(window)));
  }

  NativePointer surfacePointer() {
    return NativePointer.ofAddress(gles ? eglSurface : hdc);
  }

  void makeCurrent() {
    if (closed) {
      throw new IllegalStateException("OpenGL context is closed");
    }
    glfwMakeContextCurrent(window);
    if (gles) {
      GLES.createCapabilities();
    } else {
      GL.createCapabilities();
    }
  }

  void swapBuffers() {
    glfwSwapBuffers(window);
  }

  @Override
  public void resize(Viewport viewport) {
    makeCurrent();
  }

  @Override
  public void close() {
    if (closed) {
      return;
    }
    closed = true;
    if (window != NULL) {
      glfwMakeContextCurrent(window);
      if (gles) {
        org.lwjgl.opengles.GLES20.glFinish();
        GLES.setCapabilities(null);
      } else {
        org.lwjgl.opengl.GL11.glFinish();
        GL.setCapabilities(null);
      }
      glfwMakeContextCurrent(NULL);
      if (!gles && hdc != NULL) {
        releaseDc(glfwGetWin32Window(window), hdc);
        hdc = NULL;
      }
      glfwDestroyWindow(window);
    }
    glfwTerminate();
  }

  private static boolean isLinux() {
    return System.getProperty("os.name").toLowerCase(Locale.ROOT).contains("linux");
  }

  private static boolean isWindows() {
    return System.getProperty("os.name").toLowerCase(Locale.ROOT).contains("windows");
  }

  private static long getDc(long hwnd) {
    return User32.GetDC(hwnd);
  }

  private static void releaseDc(long hwnd, long hdc) {
    if (!User32.ReleaseDC(hwnd, hdc)) {
      throw new IllegalStateException("ReleaseDC failed");
    }
  }
}

File: examples/lwjgl-map/src/main/kotlin/org/maplibre/nativeffi/examples/lwjglmap/OpenGLContext.kt
package org.maplibre.nativeffi.examples.lwjglmap

import org.lwjgl.glfw.GLFW.GLFW_CLIENT_API
import org.lwjgl.glfw.GLFW.GLFW_CONTEXT_CREATION_API
import org.lwjgl.glfw.GLFW.GLFW_CONTEXT_VERSION_MAJOR
import org.lwjgl.glfw.GLFW.GLFW_CONTEXT_VERSION_MINOR
import org.lwjgl.glfw.GLFW.GLFW_EGL_CONTEXT_API
import org.lwjgl.glfw.GLFW.GLFW_OPENGL_API
import org.lwjgl.glfw.GLFW.GLFW_OPENGL_ES_API
import org.lwjgl.glfw.GLFW.GLFW_RESIZABLE
import org.lwjgl.glfw.GLFW.GLFW_TRUE
import org.lwjgl.glfw.GLFW.glfwCreateWindow
import org.lwjgl.glfw.GLFW.glfwDefaultWindowHints
import org.lwjgl.glfw.GLFW.glfwDestroyWindow
import org.lwjgl.glfw.GLFW.glfwGetVersionString
import org.lwjgl.glfw.GLFW.glfwInit
import org.lwjgl.glfw.GLFW.glfwMakeContextCurrent
import org.lwjgl.glfw.GLFW.glfwSwapBuffers
import org.lwjgl.glfw.GLFW.glfwTerminate
import org.lwjgl.glfw.GLFW.glfwWindowHint
import org.lwjgl.glfw.GLFWNativeEGL.glfwGetEGLConfig
import org.lwjgl.glfw.GLFWNativeEGL.glfwGetEGLContext
import org.lwjgl.glfw.GLFWNativeEGL.glfwGetEGLDisplay
import org.lwjgl.glfw.GLFWNativeEGL.glfwGetEGLSurface
import org.lwjgl.glfw.GLFWNativeWGL.glfwGetWGLContext
import org.lwjgl.glfw.GLFWNativeWin32.glfwGetWin32Window
import org.lwjgl.opengl.GL
import org.lwjgl.opengl.GL11
import org.lwjgl.opengles.GLES
import org.lwjgl.opengles.GLES20
import org.lwjgl.system.MemoryStack
import org.lwjgl.system.MemoryUtil.NULL
import org.lwjgl.system.windows.User32
import org.maplibre.nativeffi.Maplibre
import org.maplibre.nativeffi.render.OpenGLContextProvider
import org.maplibre.nativeffi.render.RenderBackend

internal class OpenGLContext private constructor(val isGles: Boolean, private val window: Long) :
  GraphicsContext {
  private var hdc = NULL
  private var eglDisplay = NULL
  private var eglConfig = NULL
  private var eglSurface = NULL
  private var eglContext = NULL
  private var closed = false

  override fun window(): Long = window

  override fun backend(): RenderBackend = RenderBackend.OPENGL

  fun hdcAddress(): Long = hdc

  fun eglDisplayAddress(): Long = eglDisplay

  fun eglConfigAddress(): Long = eglConfig

  fun eglContextAddress(): Long = eglContext

  fun wglContextAddress(): Long = glfwGetWGLContext(window)

  fun surfaceAddress(): Long = if (isGles) eglSurface else hdc

  fun makeCurrent() {
    check(!closed) { "OpenGL context is closed" }
    glfwMakeContextCurrent(window)
    if (isGles) {
      GLES.createCapabilities()
    } else {
      GL.createCapabilities()
    }
  }

  fun swapBuffers() {
    glfwSwapBuffers(window)
  }

  override fun resize(viewport: Viewport) {
    makeCurrent()
  }

  override fun close() {
    if (closed) {
      return
    }
    closed = true
    if (window != NULL) {
      glfwMakeContextCurrent(window)
      if (isGles) {
        GLES20.glFinish()
        GLES.setCapabilities(null)
      } else {
        GL11.glFinish()
        GL.setCapabilities(null)
      }
      glfwMakeContextCurrent(NULL)
      if (!isGles && hdc != NULL) {
        releaseDc(glfwGetWin32Window(window), hdc)
        hdc = NULL
      }
      glfwDestroyWindow(window)
    }
    glfwTerminate()
  }

  internal companion object {
    fun create(title: String, width: Int, height: Int): OpenGLContext {
      val providers = Maplibre.supportedOpenGLContextProviders()
      return when {
        OpenGLContextProvider.EGL in providers -> createEgl(title, width, height)
        OpenGLContextProvider.WGL in providers -> createWgl(title, width, height)
        else ->
          error(
            "The loaded MapLibre native library does not support an OpenGL context provider usable by lwjgl-map"
          )
      }
    }

    private fun createEgl(title: String, width: Int, height: Int): OpenGLContext {
      check(OpenGLContextProvider.EGL in Maplibre.supportedOpenGLContextProviders()) {
        "Native library does not support EGL"
      }
      check(glfwInit()) { "GLFW initialization failed" }
      var window = NULL
      var context = OpenGLContext(isGles = true, window = NULL)
      try {
        MemoryStack.stackPush().use { stack ->
          glfwDefaultWindowHints()
          glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_ES_API)
          glfwWindowHint(GLFW_CONTEXT_CREATION_API, GLFW_EGL_CONTEXT_API)
          glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3)
          glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0)
          glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE)
          window = glfwCreateWindow(width, height, title, NULL, NULL)
          check(window != NULL) { "GLFW EGL window creation failed" }
          context = OpenGLContext(isGles = true, window = window)
          context.makeCurrent()
          GLES.createCapabilities()
          context.eglDisplay = glfwGetEGLDisplay()
          context.eglContext = glfwGetEGLContext(window)
          context.eglSurface = glfwGetEGLSurface(window)
          val configs = stack.mallocPointer(1)
          check(glfwGetEGLConfig(window, configs)) { "GLFW did not expose an EGL config" }
          context.eglConfig = configs[0]
          check(
            context.eglDisplay != NULL && context.eglContext != NULL && context.eglSurface != NULL
          ) {
            "GLFW did not expose EGL handles"
          }
          System.out.printf("GLFW %s, OpenGL EGL/GLES%n", glfwGetVersionString())
          return context
        }
      } catch (error: RuntimeException) {
        context.close()
        if (window != NULL && context.window != window) {
          glfwDestroyWindow(window)
          glfwTerminate()
        }
        throw error
      }
    }

    private fun createWgl(title: String, width: Int, height: Int): OpenGLContext {
      check(OpenGLContextProvider.WGL in Maplibre.supportedOpenGLContextProviders()) {
        "Native library does not support WGL"
      }
      check(glfwInit()) { "GLFW initialization failed" }
      var context = OpenGLContext(isGles = false, window = NULL)
      try {
        glfwDefaultWindowHints()
        glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_API)
        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3)
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0)
        glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE)
        val window = glfwCreateWindow(width, height, title, NULL, NULL)
        check(window != NULL) { "GLFW WGL window creation failed" }
        context = OpenGLContext(isGles = false, window = window)
        context.makeCurrent()
        GL.createCapabilities()
        val hwnd = glfwGetWin32Window(window)
        val hglrc = glfwGetWGLContext(window)
        context.hdc = getDc(hwnd)
        check(context.hdc != NULL && hglrc != NULL) { "GLFW did not expose WGL handles" }
        System.out.printf("GLFW %s, OpenGL WGL%n", glfwGetVersionString())
        return context
      } catch (error: RuntimeException) {
        context.close()
        throw error
      }
    }

    private fun getDc(hwnd: Long): Long = User32.GetDC(hwnd)

    private fun releaseDc(hwnd: Long, hdc: Long) {
      check(User32.ReleaseDC(hwnd, hdc)) { "ReleaseDC failed" }
    }
  }
}
